packages/kas/src/nodes.ts

```diff
@@ -69,26 +54,77 @@ var randomFloat = function (min, max) {
 var ITERATIONS = 12;
 var TOLERANCE = 9; // decimal places
 
+// NOTE(kevinb): _.partition exists in a more recent version of underscore.
+// To avoid having to update underscore I've added a hacky version of this
+// method here.
+function partition<T, V extends _.Collection<T>>(
+    list: V,
+    iteratee: _.CollectionIterator<T, boolean>,
+): [_.TypeOfCollection<V>[], _.TypeOfCollection<V>[]] {
+    const a: _.TypeOfCollection<V>[] = [];
+    const b: _.TypeOfCollection<V>[] = [];
+    _.forEach(list, (elem, key, ctx) => {
+        if (iteratee(elem, key, ctx)) {
+            a.push(elem);
+        } else {
+            b.push(elem);
+        }
+    });
+    return [a, b];
+}
+
+function isExpr(arg: string | Expr): arg is Expr {
```

Comment:
Not a blocker, just a thought: would it be worth having these as like static methods? Like `Expr.isInstance`? Then we could make it abstract and force children to implement it; that way each variation would have a builtin predicate? idk, maybe more trouble than its worth, I just noticed the `expr instanceof Mul` below which made me think of it.

Reply:
I'm not sure if type predicates can be generic.  I'll have to experiment with that.